File: src/data/types/backgroundResult.ts
import type { CrewMember } from "./crewMember.ts";
import type { CraftResult } from "./craftResult.ts";

// Necessary completion: Agent 16's contract names
// `resolveBackgroundCrafting(...): BackgroundResult` but never defines it.
// `lastCheckedAt` advances to currentTime in both cases -- "checking on"
// a crew member resets the clock regardless of whether the background
// rate is available yet (§2.1a).
export interface BackgroundResolved {
  resolved: true;
  unitsCompleted: number;
  results: CraftResult[];
  updatedCrewMember: CrewMember;
}

export interface BackgroundRateUnavailable {
  resolved: false;
  reason: string;
  updatedCrewMember: CrewMember;
}

export type BackgroundResult = BackgroundResolved | BackgroundRateUnavailable;

File: unity/ProfitableCore/Schema/BackgroundResult.cs
namespace Profitable.Core.Schema;

// Ports src/data/types/backgroundResult.ts's BackgroundResult =
// BackgroundResolved | BackgroundRateUnavailable discriminated union.
// UpdatedCrewMember advances LastCheckedAt to currentTime in both cases --
// "checking on" a crew member resets the clock regardless of whether the
// background rate is available.
public abstract class BackgroundResult
{
    public abstract bool Resolved { get; }
}

public sealed class BackgroundResolved : BackgroundResult
{
    public override bool Resolved => true;
    public int UnitsCompleted { get; init; }
    public List<CraftResult> Results { get; init; } = new();
    public CrewMember UpdatedCrewMember { get; init; } = new();
}

public sealed class BackgroundRateUnavailable : BackgroundResult
{
    public override bool Resolved => false;
    public string Reason { get; init; } = string.Empty;
    public CrewMember UpdatedCrewMember { get; init; } = new();
}
